javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -758,21 +774,22 @@ public String toString() {
          * *     let val x = findMin ts
          * *     in if Elem.leq (root t, x) then root t else x end
          */
-        protected static <T> T findMin(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest) {
-            if (forest.isEmpty()) {
-                throw new NoSuchElementException();
-            } else {
-                final T value = root(forest.head());
+        protected static <T> T findMin(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest, T currentMin) {
+            while (true) {
```

Comment:
another tailrec unrolling by separating it into two methods (it's also a lot simpler to understand)